### PR TITLE
Various SQL export additions

### PR DESCRIPTION
**Pull request by NoahCarnahan**
Best reviewed commit-by-commit
@benrudolph @snopoke cc @kaapstorm

### CODE REVIEW

File: corehq/apps/export/export.py [hunk 1]
```diff
@@ -61,16 +70,50 @@ def get_preview(self):
...
+    """
+    format = Format.XLS_2007
+    if len(export_instances) == 1:
+        format = export_instances[0].export_format
+
+    legacy_writer = get_writer(format)
+    writer = _Writer(legacy_writer)
+    return writer
+
+def _get_tables(export_instances):
```
Comment (dimagimon) [inline corehq/apps/export/export.py]: E302 expected 2 blank lines, found 1

File: corehq/apps/export/export.py [hunk 2]
```diff
@@ -80,21 +123,27 @@ def _get_export_documents(export_instance, filters):
...
-        for doc in documents:
-            for table in export_instance.tables:
-                rows = table.get_rows(doc)
-                for row in rows:
-                    # It might be bad to write one row at a time when you can do more (from a performance perspective)
-                    # Regardless, we should handle the batching of rows in the _Writer class, not here.
-                    writer.write(table, row)
 
-    return ExportFile(writer.path)
+def _write_export_instance(writer, export_instance, documents):
```
Comment (dimagimon) [inline corehq/apps/export/export.py]: E303 too many blank lines (5)

File: corehq/apps/export/tests/test_get_export_file.py [hunk 3]
```diff
@@ -209,3 +295,86 @@ def test_get_export_file(self):
...
+                        wb[sheet][cell].value,
+                        expected[sheet][cell],
+                        'AssertionError: Sheet "{}", cell "{}" expected: "{}", got "{}"'.format(
+                            sheet, cell, expected[sheet][cell], wb[sheet][cell].value
+                        )
+                    )
+
+
+
+class TableHeaderTest(SimpleTestCase):
```
Comment (dimagimon) [inline corehq/apps/export/tests/test_get_export_file.py]: E303 too many blank lines (3)

File: corehq/apps/export/const.py [hunk 4]
```diff
@@ -1,3 +1,7 @@
+"""
+Some of these constants need correspond to constants set in corehq/apps/export/static/export/js/const.js
```
Comment (snopoke) [inline corehq/apps/export/const.py]: Typo: `need`

File: corehq/apps/export/esaccessors.py [hunk 5]
```diff
@@ -10,8 +10,8 @@ def get_form_export_base_query(domain, app_id, xmlns, include_errors):
             .sort("received_on"))
     if include_errors:
         query = query.remove_default_filter("is_xform_instance")
-        # Assuming that the only other things in the FormES index are form errors,
-        # but perhaps I should filter on those types explicitly.
+        query = query.doc_type(["xforminstance", "xformarchived", "xformdeprecated", "xformduplicate"])
+        # TODO: Confirm that the doc types get lower cased like this
```
Comment (snopoke) [inline corehq/apps/export/esaccessors.py]: They do but I think ES will do a lowercase comparison anyway

File: corehq/apps/export/const.py [hunk 6]
```diff
@@ -27,10 +31,12 @@
 FORM_EXPORT = 'form'
 CASE_EXPORT = 'case'
 
-MAIN_TABLE = None
+MAIN_TABLE = []
```
Comment (snopoke) [inline corehq/apps/export/const.py]: Why the change to `[]`

File: corehq/apps/export/models/new.py [hunk 7]
```diff
@@ -84,7 +84,7 @@ class ExportColumn(DocumentSchema):
     # A list of constants that map to functions to transform the column value
     transforms = ListProperty(validators=is_valid_transform)
 
-    def get_value(self, doc, base_path):
+    def get_value(self, doc, base_path, transform_dates=False):
```
Comment (snopoke) [inline corehq/apps/export/models/new.py]: We seem to be going back and forth on the `transform_dates` field

File: corehq/apps/export/models/new.py [hunk 8]
```diff
@@ -153,8 +157,18 @@ def update_properties_from_app_ids_and_versions(self, app_ids_and_versions):
...
+    def is_deidentifed(self):
+        # TODO: Might be better if we set an is_deidentified flag on the model instead?
+        return bool(set(self.transforms) & set(DEID_TRANSFORM_FUNCTIONS))
+
     def get_headers(self):
-        return [self.label]
+        # TODO: id columns need special treatment
+        # see couchexport.models.ExportTable#get_headers_row
+        if self.is_deidentifed:
+            return [u"{}{}".format(self.label, "[sensitive]")]
```
Comment (snopoke) [inline corehq/apps/export/models/new.py]: Should there be a space between the label and the `[sensitive]` tag?

**Comment by snopoke** [comment]
looking good, will let @benrudolph review too

File: corehq/apps/export/const.py (hunk 6 above)
Comment (NoahCarnahan) [inline corehq/apps/export/const.py]: Sorry, should have explained. It's follow up from here: https://github.com/dimagi/commcare-hq/pull/10182#discussion_r51935186

File: corehq/apps/export/models/new.py (hunk 7 above)
Comment (NoahCarnahan) [inline corehq/apps/export/models/new.py]: Yeah. I added it back to conform to Ben's latest PR: https://github.com/dimagi/commcare-hq/pull/10318

File: corehq/apps/export/esaccessors.py (hunk 5 above)
Comment (NoahCarnahan) [inline corehq/apps/export/esaccessors.py]: :+1:

File: corehq/apps/export/models/new.py (hunk 8 above)
Comment (NoahCarnahan) [inline corehq/apps/export/models/new.py]: woops

File: corehq/apps/export/const.py (hunk 6 above)
Comment (snopoke) [inline corehq/apps/export/const.py]: 👍

On 17 February 2016 at 15:37, Noah Carnahan <notifications@github.com>
wrote:

> In corehq/apps/export/const.py
> <https://github.com/dimagi/commcare-hq/pull/10322#discussion_r53163691>:
>
> > @@ -27,10 +31,12 @@
> >  FORM_EXPORT = 'form'
> >  CASE_EXPORT = 'case'
> >
> > -MAIN_TABLE = None
> > +MAIN_TABLE = []
>
> Sorry, should have explained. It's follow up from here: #10182 (comment)
> <https://github.com/dimagi/commcare-hq/pull/10182#discussion_r51935186>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/10322/files#r53163691>.
>



-- 
Simon Kelly
Senior Engineer | Dimagi South Africa

File: corehq/apps/export/models/new.py (hunk 7 above)
Comment (snopoke) [inline corehq/apps/export/models/new.py]: 👍

On 17 February 2016 at 15:40, Noah Carnahan <notifications@github.com>
wrote:

> In corehq/apps/export/models/new.py
> <https://github.com/dimagi/commcare-hq/pull/10322#discussion_r53164013>:
>
> > @@ -84,7 +84,7 @@ class ExportColumn(DocumentSchema):
> >      # A list of constants that map to functions to transform the column value
> >      transforms = ListProperty(validators=is_valid_transform)
> >
> > -    def get_value(self, doc, base_path):
> > +    def get_value(self, doc, base_path, transform_dates=False):
>
> Yeah. I added it back to conform to Ben's latest PR: #10318
> <https://github.com/dimagi/commcare-hq/pull/10318>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/10322/files#r53164013>.
>



-- 
Simon Kelly
Senior Engineer | Dimagi South Africa

**Comment by NoahCarnahan** [comment]
Thanks for review @snopoke. Just pushed a couple commits addressing feedback.

**Comment by benrudolph** [comment]
looks good! tests are broken, but after that g2m

**Comment by NoahCarnahan** [comment]
:white_check_mark:
